More granular request process tracking

kazimuth opened the pull request:
## Description of Changes
Addresses https://github.com/clockworklabs/SpacetimeDBPrivate/issues/1786 and https://github.com/clockworklabs/com.clockworklabs.spacetimedbsdk/issues/329. Marks messages as parsed from the background parsing thread, and then marks them as applied using a separate tracker once they are applied.

## API

 - [ ] This is an API breaking change to the SDK

*If the API is breaking, please state below what will break*

## Requires SpacetimeDB PRs

## Testsuite

SpacetimeDB branch name: master

## Testing

- [x] Testing against Bitcraft

kazimuth committed:
More granular request process tracking

kazimuth committed:
Docs on trackers + invoke trackers on background thread

kazimuth commented:
I've tested this against BitCraft locally, and opened https://github.com/clockworklabs/BitCraft/pull/5506 which can be merged after this PR is merged and released.

kazimuth commented:
![Screenshot 2025-06-12 165946](https://github.com/user-attachments/assets/08b459c8-6949-47eb-8aac-abadeb6cf522)

It looks like messages wait in the queue to be applied a lot longer than they wait in the queue to be parsed, in the un-loaded case.

kazimuth commented:
Note that this changes "Reducer Latency" from measuring `call -> apply` time to measuring `call -> receive` time. The order of events is

```
call    ->      receive      ->      parse     ->    apply
     network               client            client
```

kazimuth committed:
Remove dead import

joshua-spacetime commented on src/Stats.cs:
```diff
@@ -229,10 +235,41 @@ internal void InsertRequest(DateTime start, string metadata)
     {
+        /// <summary>
+        /// Tracks times from reducers requests being sent to their responses being received.
+        /// This is clientside time, not host-time: for that you want AllReducersTracker.
```
This is `host-time + network + outer parse-time`?

kazimuth replied:
Yes, I added comments clarifying this.
I could break it down further with measurements of time waiting in queue versus actually being parsed. Do you think that would be helpful?

joshua-spacetime commented on src/Stats.cs:
```diff
@@ -229,10 +235,41 @@ internal void InsertRequest(DateTime start, string metadata)
         public readonly NetworkRequestTracker ReducerRequestTracker = new();
+
+        /// <summary>
+        /// Tracks times from one-off requests being sent to their responses being received.
```
This does not include message parsing/processing? Does it include any time waiting in queue?

kazimuth replied:
This includes queue and parsing times because OneOffQuery is somewhat weird.

joshua-spacetime commented on src/Stats.cs:
```diff
@@ -229,10 +235,41 @@ internal void InsertRequest(DateTime start, string metadata)
+        /// <summary>
+        /// Tracks times from subscriptions being sent to their responses being received.
+        /// Includes the time needed to parse the outer layer of the response message, but not the inner layer
+        /// if it includes a DatabaseUpdate.
```
Does it include time waiting in queue?

kazimuth replied:
Yes, clarified comment.

joshua-spacetime commented on src/Stats.cs:
```diff
@@ -229,10 +235,41 @@ internal void InsertRequest(DateTime start, string metadata)
         public readonly NetworkRequestTracker ParseMessageTracker = new();
+
+        /// <summary>
+        /// Tracks times from messages being parsed on a background thread to their being applied on the main thread.
```
Does this one include time waiting in queue?

kazimuth replied:
Yes, clarified comment.

kazimuth committed:
Clearer comments

kazimuth committed:
More granular trackers, and refine names in SpacetimeDBClient.cs to match those in Stats.cs

kazimuth commented:
I've split the trackers to be more granular, and updated SpacetimeDBClient's internal names to match the names we're using in `Stats.cs`. This should make the code easier to understand; it's not a user-visible change.

joshua-spacetime reviewed (approved):
Approving, but with one request. Can you further clarify the difference between `ParseMessageQueueTracker` and `ParseMessageTracker`? Specifically which one includes parse time and which one does not, assuming both include the queue time.

rekhoff reviewed (approved):
These changes look great. The code is well commented and provides greater details about how the metrics are calculated. My only request as a follow-up might be to add on the BitCraft side, some timing descriptions to the F9 menu text, so what timing specifically is being shown comes just after the name of the metric. I'll approve this as well.

kazimuth committed:
Fix comments

kazimuth commented:
> Approving, but with one request. Can you further clarify the difference between `ParseMessageQueueTracker` and `ParseMessageTracker`? Specifically which one includes parse time and which one does not, assuming both include the queue time.

One of them is only queue time, and one is only parse time.
